### PR TITLE
Enable custom detailed PV layout models

### PR DESCRIPTION
**Pull request by Matthew-Boyd**
Enables custom detailed PV layout models for both the detailed (pvsamv1) and simple (pvwattsv8) technology models.

### CODE REVIEW

**Comment by Matthew-Boyd** [comment]
@dguittet 

- should detailed_pv_layout.py and detailed_pv_config.py be developed more before merging into master?
- or should they maybe be located somewhere else, like in /examples?

**Comment by dguittet** [comment]
@Matthew-Boyd Apologies for the delay. Looking at this now. I don't think detailed_pv_layout.py and detailed_pv_config.py were intended to be merged into main HOPP since they're not complete. Do you think it makes sense to not include them in the main HOPP or would something be lacking? Examples could be a good idea!

File: hybrid/layout/pv_design_utils.py [hunk 1]
```diff
@@ -210,7 +237,7 @@ def spe_power(spe_eff_level, spe_rad_level, spe_area) -> float:
     return spe_eff_level / 100 * spe_rad_level * spe_area
 
 
-def get_module_attribs(pvsam_model: pv.Pvsamv1) -> dict:
+def get_module_attribs(pvsam_model: pv_detailed.Pvsamv1) -> dict:
```
Comment (dguittet) [inline hybrid/layout/pv_design_utils.py:240]: Would be really nice to have a docstring here linking to the attributes in the PySAM documentation

File: hybrid/layout/pv_design_utils.py [hunk 2]
```diff
@@ -272,14 +299,14 @@ def get_module_attribs(pvsam_model: pv.Pvsamv1) -> dict:
             'length':       module_length,
             'I_mp_ref':     I_mp,
             'I_sc_ref':     I_sc,
-            'P_mp_ref':     P_mp,
+            'P_mp_ref':     P_mp * 1e-3,
```
Comment (dguittet) [inline hybrid/layout/pv_design_utils.py:302]: could you add units to these key-value pairs too as you did below?

File: hybrid/layout/pv_design_utils.py [hunk 3]
```diff
@@ -272,14 +299,14 @@ def get_module_attribs(pvsam_model: pv.Pvsamv1) -> dict:
...
             'V_mp_ref':     V_mp,
             'V_oc_ref':     V_oc,
             'width':        module_width
         }
 
 
-def get_inverter_attribs(pvsam_model: pv.Pvsamv1) -> dict:
+def get_inverter_attribs(pvsam_model: pv_detailed.Pvsamv1) -> dict:
```
Comment (dguittet) [inline hybrid/layout/pv_design_utils.py:309]: Would be really nice to have a docstring here linking to the attributes in the PySAM documentation

File: hybrid/pv_source.py [hunk 4]
```diff
@@ -35,10 +35,15 @@ def __init__(self,
 
         self.dc_degradation = [0]
 
-        params: Optional[PVGridParameters] = None
-        if 'layout_params' in pv_config.keys():
-            params: PVGridParameters = pv_config['layout_params']
-        self._layout = PVLayout(site, system_model, params)
+        if 'layout_model' in pv_config.keys():
```
Comment (dguittet) [inline hybrid/pv_source.py:38]: Can this `layout_model` key-value pair be documented in this file?

File: hybrid/layout/pv_layout.py [hunk 5]
```diff
@@ -52,10 +52,10 @@ def __init__(self,
...
         self.min_spacing = min_spacing
 
-        self.module_power: float = module_power
-        self.module_width: float = module_width
-        self.module_height: float = module_height
-        self.modules_per_string: int = modules_per_string
+        self.module_power: float = get_module_power(self._system_model)
+        self.module_width: float = get_module_width(self._system_model)
+        self.module_height: float = get_module_length(self._system_model)
+        self.modules_per_string: int = get_modules_per_string(self._system_model)
```
Comment (dguittet) [inline hybrid/layout/pv_layout.py:58]: this seems less efficient than calling `get_module_attribs` and assigning from the dictionary

File: hybrid/layout/pv_layout.py [hunk 6]
```diff
@@ -213,8 +213,8 @@ def set_system_capacity(self,
         """
         Changes system capacity in the existing layout
         """
-        if type(self.parameters) == PVGridParameters:
-            self.reset_solargrid(size_kw, self.parameters)
+        if type(self.parameters) == PVGridParameters or "DetailedPVParameters" in str(type(self.parameters)):
```
Comment (dguittet) [inline hybrid/layout/pv_layout.py:216]: If someone named their parameters differently, then this code wouldn't be flexible enough to handle it. I think we should throw an error to stave off confusion

**Review by dguittet — COMMENTED** [review]
I'm really liking the capability to pass in instantiated component classes like layout.

File: hybrid/layout/pv_design_utils.py (hunk 2 above)
Comment (Matthew-Boyd) [reply hybrid/layout/pv_design_utils.py:302]: Done

File: hybrid/pv_source.py (hunk 4 above)
Comment (Matthew-Boyd) [reply hybrid/pv_source.py:38]: Done

File: hybrid/layout/pv_layout.py (hunk 5 above)
Comment (Matthew-Boyd) [reply hybrid/layout/pv_layout.py:58]: `get_module_attribs` is specific to pvsamv1 where these functions provide parameters for either pvsamv1 or pvwatts, where the pvwatts parameters come from `hybrid.layout.pv_module`. You're right though--I've moved this functionality into `get_module_attribs`.

File: hybrid/layout/pv_design_utils.py (hunk 1 above)
Comment (Matthew-Boyd) [reply hybrid/layout/pv_design_utils.py:240]: Done

File: hybrid/layout/pv_design_utils.py (hunk 3 above)
Comment (Matthew-Boyd) [reply hybrid/layout/pv_design_utils.py:309]: Done

File: hybrid/layout/pv_layout.py (hunk 6 above)
Comment (Matthew-Boyd) [reply hybrid/layout/pv_layout.py]: I'm not sure now why I even needed to accommodate the `DetailedPVParameters` type, so I just removed it. I think the tests will pass, we'll see.

**Comment by Matthew-Boyd** [comment]
> @Matthew-Boyd Apologies for the delay. Looking at this now. I don't think detailed_pv_layout.py and detailed_pv_config.py were intended to be merged into main HOPP since they're not complete. Do you think it makes sense to not include them in the main HOPP or would something be lacking? Examples could be a good idea, but I think it would need to be more functional / complete.

@dguittet No worries, I got to catch up on other work. I'm content putting them into examples as-is now, and let Engie use this right away. Maybe they would offer comments or a PR during their testing and implementation? I'd rather not let the changes in this PR languish, however, so if you thought it best to develop those two files more, maybe we can still merge the rest? (I would have to remove the affected tests, though.)

**Comment by dguittet** [comment]
@Matthew-Boyd Thanks I've moved the detailed layout files into the examples folder

**Comment by Matthew-Boyd** [comment]
@dguittet Good to merge?

File: examples/Detailed_PV_Layout/detailed_pv_config.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        e.g. `PVLayoutConfig(**layout_config_dict)`
+
+    This dataclass will be used to construct the DetailedPVPlant, and passed along to DetailedPVLayout
+
+    Module and inverter data is not stored here, but this could be changed
+
+    Any pvsamv1 variables besides module and inverter variables that are not listed here will remain unassigned in the yield model. 
+    Some pvsamv1 variables have default-if-not-assigned values, while others are required and will cause an exception if not provided.
+    The former kind of variable should be checked to see if they are worth adding to this class, while the
+    second kind should be added to this class.
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply examples/Detailed_PV_Layout/detailed_pv_config.py]: Done

File: examples/Detailed_PV_Layout/detailed_pv_layout.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    and module-inverter counts for a single subarray. The design can be expanded to include rack configs and locations, 
+    a set of wires, any losses, other power electronics, etc.
+
+    The PV layout design is calculated using the inputs from PVLayoutConfig and from the design vector DetailedPVParameters
+    so that the layout best matches the target design vector
+
+    Layout is not computed during construction, but happens during these function calls:
+        1. `compute_pv_layout`: this uses the existing DetailedPVParameters
+        2. `set_layout_params`: this uses new DetailedPVParameters
+
```
[the remark this answers is not in the document]
Comment (Matthew-Boyd) [reply examples/Detailed_PV_Layout/detailed_pv_layout.py:43]: Done

**Review by dguittet — APPROVED** [review]

**Pull request merged**